File: 02-week-2--recursion/05-day-5--recursive-selection-sort/ruby/selection_sort_recursive.rb
def selection_sort_recursive(arr)
  # type your code in here
  # base case if array is empty
  if arr.empty?
    return []
  end

  # get the smallest number in the array
  minimum = arr.min
  # and the index of the FIRST occurrence
  index = arr.index(minimum)
  # delete it from the input array and return that value
  # arr.delete_at(index)

  ## delete removes duplicates, so that doesnt work
  ## arr.delete(minimum)

  # create a new array to hold the sorted values 
  sorted = selection_sort_recursive(arr)
  # push each minimum value into the sorted array until the input array is empty
  sorted.unshift(minimum)
  
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: [-1, 2, 3, 5]"
  print "=> "
  print selection_sort_recursive([3, -1, 5, 2])

  puts

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: 02-week-2--recursion/05-day-5--recursive-selection-sort/javascript/selection_sort_recursive.js
function selectionSortRecursive(arr) {
  // type your code here
  // output a new array from the input
  // specifically wants an empty array to be returned if the array is empty
  if (arr.length == 0){
    return []
  }
  // math min returns the smallest element in a given input, but not as an array
  let minimum = Math.min(...arr)
  // get that index number from the input array
  let index = arr.indexOf(minimum)
  // once you have that number, remove it from the input array
  arr.splice(index, 1)
  
  // test will be our output array
  let test = selectionSortRecursive(arr)
  // minimum gets moved into the new array
  test.unshift(minimum)
  // keep running test until the arr array is empty
  return test
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: [-1, 2, 3, 5]");
  console.log("=>", selectionSortRecursive([3, -1, 5, 2]));

  console.log("");
}

module.exports = selectionSortRecursive;

// Please add your pseudocode to this file
// And a written explanation of your solution
